libc: minimal: Add strtoll() and strtoull()

MeisterBob opened the pull request:
strtoll() and strtoull() are copies of strtol() and strtoul() with types
changed to long long instead of long.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

Closes #36645

stephanosio reviewed (changes requested):
1. Please add the tests for these functions in `tests/lib/c_lib`.
2. Need to update the relevant documentation:
https://github.com/zephyrproject-rtos/zephyr/blame/main/doc/guides/portability/posix.rst#L311
3. Need to remove the local `strtoll` implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/samples/net/civetweb/common/src/libc_extensions.c#L156
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/samples/net/civetweb/common/include/libc_extensions.h#L29

mt-gjoerges committed:
libc: minimal: Add strtoll() and strtoull()

- strtoll() and strtoull() are copies of strtol() and strtoul() with
  types changed to long long instead of long.
- added tests
- added documentation
- removed stubs from civetweb sample

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

MeisterBob commented:
I was busy in an other project the last weeks, but finally I added tests and documentation and removed stubs from civetweb samples. Also rebased the branch..

MeisterBob commented:
I stumbled across https://github.com/zephyrproject-rtos/zephyr/pull/16511 by @PiotrZierhoffer. I liek the way he unified the conversion mechanism. The only thing missing in his PR where tests. I would suggest to take the tests I implemented and put them into his PR.

rivimey commented on tests/lib/c_lib/src/main.c:
```diff
@@ -812,6 +812,145 @@ void test_strtoul(void)
+
+	char border1[] = "+18446744073709551615";
+	char border2[] = "-18446744073709551615000";
+	char border3[] = "+18446744073709551619";
```
Suggest adding a couple more cases here to test a long string of digits but not a large number, and a simply very large number:
```
border[] = "0x0000000000000000000000000000000000001"; // should pass
border[] = "10000000000000000000000000000000000001"; // should fail (too big)
border[] = "-10000000000000000000000000000000000001"; // should fail (too big)
```

MeisterBob replied:
added those cases to strtoll and strtoull test.

tejlmand reviewed (approved):
lgtm - reapprove

stephanosio reviewed (changes requested):
This PR contains the source files that are `BSD-4-Clause-UC` licensed.

This is different to the usual 3-clause BSD license variants that we have previously accepted into the codebase; in particular, the third clause is likely going to be a problem for the project.

```
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgment:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
```

If we are to accept this as is, this needs to go through the board as per the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#components-using-other-licenses).

The most practical solution would be to replace the strto(u)ll implementation with the one that is more freely licensed (e.g. BSD-2-Clause, BSD-3-Clause, ...).

cc @nashif @carlescufi

MeisterBob commented:
It's the same license [strtol](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtol.c) and  [strtoul](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtoul.c) are using. If you can point out an implementation that more suitable licensed I would be happy to replace strtol, strtoul, strtoll and strtoull.

stephanosio commented:
> It's the same license [strtol](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtol.c) and [strtoul](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtoul.c) are using. If you can point out an implementation that more suitable licensed I would be happy to replace strtol, strtoul, strtoll and strtoull.

@MeisterBob You are right. They are indeed `BSD-4-Clause-UC` (not sure why these did not turn up while grepping ...).

In that case, I assume this license has already been approved by the board. I would like @nashif to give a confirmation though, since those date back to the beginning of the project (maybe before such rules were in place?) ...

carlescufi reviewed (changes requested):
I am lookin at adding a few more of those. Can you please move those to a single source file and combine the use of the whitespace trimming so that it is not duplicated for every function?

carlescufi reviewed (approved):
>I am lookin at adding a few more of those. Can you please move those to a single source file and combine the use of the whitespace trimming so that it is not duplicated for every function?

Please ignore this, I will take care of this in a follow-up PR.

de-nordic commented:
Shouldn't this addition be included into release notes?

carlescufi commented:
This replaces #43289 for the mesh code, with further additions that I will provide.

carlescufi commented:
> Shouldn't this addition be included into release notes?

Release notes are populated at release time generally, except for API breakage/changes.

carlescufi commented:
Approved by the TSC, merging with failing checks:

- Coding guidelines: Use of reserved identifiers (this is the C library)
- Scancode: Use of a non-Apache license (approved by the TSC)

kestewart commented:
This is a non-OSI approved license.  Per:https://docs.zephyrproject.org/latest/contribute/external.html#external-src-process it's going to need board review.
